### ISSUE TITLE
Non-unique files/folders written in temp directory (analysis runs fail with multiple users)

### ISSUE THREAD

**Issue by sampie** [biocore/qiime#1870]
I have Qiime 1.8 and Qiime 1.9rc1 in a server having multiple users. Qiime seems to write non-unique files to tmp folder (/tmp by default). For example, user A is running Qiime which writes /tmp/R.stdout file. Now, user B is running Qiime and it is trying to write /tmp/R.stdout, but it fails because the file already exists and it is owner by user A.

**Comment by jairideout** [comment]
Thanks for reporting this! What script was being run that caused this error? Does this issue happen with both 1.8.0 and 1.9.0-rc1?

**Comment by sampie** [comment]
For Qiime 1.9rc1, I can see following error:

Traceback (most recent call last):   
  File "/home/apps/qiime-env/1.9rc1/bin/compare_categories.py", line 189, in <module>
    main()
  File "/home/apps/qiime-env/1.9rc1/bin/compare_categories.py", line 185, in main
    categories, opts.num_permutations, out_dir)
  File "/home/apps/qiime-env/1.9rc1/local/lib/python2.7/site-packages/qiime/compare_categories.py", line 146, in compare_categories
    rex(command_args, '%s.r' % method, output_dir=out_dir)
  File "/home/apps/qiime-env/1.9rc1/local/lib/python2.7/site-packages/qiime/util.py", line 1965, in __call__
    outfile = open(outfilepath, 'w') 
IOError: [Errno 13] Permission denied: '/tmp/R.stdout'

-----

I am also seeing:

Traceback (most recent call last):
  File "/home/apps/qiime-env/1.9rc1/bin/parallel_align_seqs_pynast.py", line 104, in <module>
    main()
  File "/home/apps/qiime-env/1.9rc1/bin/parallel_align_seqs_pynast.py", line 100, in main
    suppress_submit_jobs=opts.suppress_submit_jobs)
  File "/home/apps/qiime-env/1.9rc1/local/lib/python2.7/site-packages/qiime/parallel/util.py", line 133, in __call__
    self._precommand_initiation(input_fp, output_dir, working_dir, params)
  File "/home/apps/qiime-env/1.9rc1/local/lib/python2.7/site-packages/qiime/parallel/align_seqs.py", line 32, in _precommand_initiation
    output_dir=get_qiime_temp_dir())
  File "/home/apps/qiime-env/1.9rc1/local/lib/python2.7/site-packages/bfillings/formatdb.py", line 121, in build_blast_db_from_fasta_path
    app_result = fdb(fasta_path)
  File "/home/apps/qiime-env/1.9rc1/local/lib/python2.7/site-packages/burrito/util.py", line 284, in __call__
    'StdErr:\n%s\n' % open(errfile).read())
burrito.util.ApplicationError: Unacceptable application exit status: 2
Command:
cd "/tmp//"; formatdb -l "core_set_aligned.fasta.imputed.log" -o T -n "core_set_aligned.fasta.imputed" -i "/home/apps/qiime-env/1.9rc1/qiime_software/core_set_aligned.fasta.imputed" -p F > "/dev/null" 2> "/dev/null"

**Comment by jairideout** [comment]
Thanks! I confirmed this is a bug, though this code hasn't changed in awhile (since long before 1.8.0). It's strange that the issue is appearing now. Do you see this with QIIME 1.8.0?

**Comment by sampie** [comment]
Thanks for taking a look at this issue! I have also seen similar permission problems with Qiime 1.8 in multiple user environment. In that sense, this is not a regression bug. However, it has been a while since I last tested 1.8.

**Comment by jairideout** [comment]
Ok thanks, we'll be sure to get this fixed for 1.9.0.

**Comment by wasade** [comment]
What's the qiime tmp directory?

**Comment by sampie** [comment]
Thank you! I'll be happy to test the fix once it becomes available.

wasade: Qiime temp is a directory for temporary files created by qiime pipeline/scripts. It is specified by temp_dir in qiime_config file.

**Comment by wasade** [comment]
Yes, what is it set to, /tmp? what I'm trying to determine is if R is
disregarding temp_dir

**Comment by jairideout** [comment]
@wasade the R code is creating `R.stdout` and `R.stderr` within the QIIME temp directory. I think the code needs to be updated to create safe temporary files within the QIIME temp directory to avoid naming conflicts between users.

**Comment by wasade** [comment]
Agree it wouldnt hurt, but users could also specify their own tmp
directory. /tmp on compute nodes isn't assured to be large anyway and on
large resources it likely is more efficient to use scratch if available

**Comment by jairideout** [comment]
It should be an easy fix and it'd be nice to not force each user to have to specify a different temp directory in their QIIME config file.

**Comment by sampie** [comment]
Wow, that was fast. Is there a way to install the fixed version of Qiime with PIP?

wasade: This test was using Qiime default /tmp. In clustered environment there is /worktemp (shared among all machines and among all users).

**Comment by jairideout** [comment]
My pull request hasn't been merged yet (https://github.com/biocore/qiime/pull/1881). I'll post back here with instructions for upgrading once it's been merged. Thanks for offering to help test this!

**Comment by sampie** [comment]
Hi Jai,

I have taken a look at Qiime code (https://github.com/biocore/qiime.git). start_parallel_jobs.py seems to write "jobs" directory to tmp_dir, which might cause also similar permission problems. I think this jobs directory name should be unique. 

    jobs_dir = '%s/jobs/' % output_dir
    job_fps = []
    if not exists(jobs_dir):
        try:
            makedirs(jobs_dir)
        except OSError as e:
            raise OSError("Error creating jobs directory in temp dir: %s" % output_dir +
                          " (specified in qiime_config). Do you have write access?. " +
                          "Original error message follows:\n%s" % str(e))

**Comment by sampie** [comment]
Here are more non-unique files written to tmp, which cause problems. The cause seems to be the known code section.

File "/home/apps/qiime-env/1.9rc2/bin/parallel_align_seqs_pynast.py", line 104, in <module>
    main()
  File "/home/apps/qiime-env/1.9rc2/bin/parallel_align_seqs_pynast.py", line 100, in main
    suppress_submit_jobs=opts.suppress_submit_jobs)
  File "/home/apps/qiime-env/1.9rc2/local/lib/python2.7/site-packages/qiime/parallel/util.py"
, line 133, in __call__
    self._precommand_initiation(input_fp, output_dir, working_dir, params)
  File "/home/apps/qiime-env/1.9rc2/local/lib/python2.7/site-packages/qiime/parallel/align_se
qs.py", line 32, in _precommand_initiation
    output_dir=get_qiime_temp_dir())
  File "/home/apps/qiime-env/1.9rc2/local/lib/python2.7/site-packages/bfillings/formatdb.py",
 line 121, in build_blast_db_from_fasta_path
    app_result = fdb(fasta_path)
  File "/home/apps/qiime-env/1.9rc2/local/lib/python2.7/site-packages/burrito/util.py", line 
284, in __call__
    'StdErr:\n%s\n' % open(errfile).read())
burrito.util.ApplicationError: Unacceptable application exit status: 2
Command:
cd "/tmp//"; formatdb -l "85_otus.fasta.log" -o T -n "85_otus.fasta" -i "/home/apps/qiime-env/1.9rc2/local/lib/python2.7/site-packages/qiime_default_reference/gg_13_8_otus/rep_set_aligned/85_otus.fasta" -p F > "/dev/null" 2> "/dev/null"

----

85_otus.fasta.log
85_otus.fasta.nhr
85_otus.fasta.nin
85_otus.fasta.nsd
85_otus.fasta.nsi
85_otus.fasta.nsq

----

core_set_aligned.fasta.imputed.log
core_set_aligned.fasta.imputed.nhr
core_set_aligned.fasta.imputed.nin
core_set_aligned.fasta.imputed.nsd
core_set_aligned.fasta.imputed.nsi
core_set_aligned.fasta.imputed.nsq

**Comment by jairideout** [comment]
Thanks @sampie! The `R.stdout`/`R.stderr` fix has been merged into master. Can you please try it out and let me know how it works for you? In the meantime I'll look into the other temp directory issues you posted.

To upgrade to the latest development version of QIIME, run:

```
pip install https://github.com/biocore/qiime/archive/master.zip --upgrade
```

Depending on how you previously installed the release candidate, you may need to prefix the command with `sudo`.

**Comment by sampie** [comment]
Hi Jai,

I tested the latest development version. I did not get any R.stdout / R.stderr errors. Nice. 

The analysis crashes to the remaining permission denied issues. I have not yet tested slurm configuration.

btw. The script I am using to find problems is available at (https://ngs.utu.fi/files/qiime/). It runs Qiime pipeline when given a folder with per sample R1 and R2 fastq files and a mapping file.

**Comment by jairideout** [comment]
Awesome, glad to hear the R stdout/stderr bug is fixed!

As @wasade suggested, do you still get errors with the other scripts if you change your QIIME temp directory to something that's unique to your user?

I'm still planning to look into these other failures, but I think they might be harder to fix because of how QIIME's parallel framework works (I think these issues have been in QIIME for awhile). Fixing this issue may also require changes in one or more of QIIME's dependencies, which means we'll need new releases of those package(s). I'll check with @gregcaporaso to see if this is feasible to fix for the 1.9.0 release and then follow up here.

**Comment by sampie** [comment]
Hi Jai,

I am not exactly sure how to configure separate temp dir for each user. QIIME_CONFIG_FP points to the global qiime_config file and it has temp_dir. I am not sure how to specify there a user specific folder like: /worktemp/$username/. Then I have thought about instructing each usersto create .qiime_config having only "temp_dir /worktemp/<username>/, but that would probably leave lots of .qiime_config files in the future to user folders which override the parameter even when it is not desired anymore.

If there an unix environment variable to override Qiime temp_dir? That kind of variable would be easy to set automatically as user specific and it can be fully controlled by system admin.

**Comment by jairideout** [comment]
What about adding the following line to your global QIIME config file (the one that's pointed at by `$QIIME_CONFIG_FP`):

```
temp_dir /worktemp/$USER
```

That's the strategy we use on our institution's cluster.

**Comment by sampie** [comment]
Hi Jai,

That sounds like a workaround I wanted but did not know how to configure. I'll test that. Thanks!

**Comment by sampie** [comment]
The workaround works. There were no errors with this configuration.

**Comment by jairideout** [comment]
Thanks for testing this! @gregcaporaso and I don't think it's feasible to fix the non-unique temp file issue in parallel QIIME for the 1.9.0 release. We'll update the [parallel QIIME tutorial](http://qiime.org/tutorials/parallel_qiime.html) to explain that a unique temporary directory is necessary when using parallel QIIME in a multi-user environment, and will also include the workaround I described above.

**Comment by sampie** [comment]
That sounds like a good solution for Qiime 1.9.0 Thanks for your help!

**Comment by jairideout** [comment]
No problem! I submitted a pull request (#1898) with updates and additional documentation in the parallel QIIME tutorial.

**Comment by sampie** [comment]
Perhaps this bug report should remain open until it is fixed, as only a workaround exists now?

**Comment by jairideout** [comment]
I think this is okay to close because we will be completely redesigning the way parallel QIIME works for QIIME 2, so this won't be an issue with the new system. This issue isn't something we'll be able to tackle for QIIME 1.x.

**Comment by sampie** [comment]
Sounds good.
